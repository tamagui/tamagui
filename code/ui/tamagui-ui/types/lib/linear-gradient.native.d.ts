export * from 'expo-linear-gradient';
//# sourceMappingURL=linear-gradient.native.d.ts.map