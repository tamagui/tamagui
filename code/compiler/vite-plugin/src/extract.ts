// fork from https://github.com/seek-oss/vanilla-extract

import type { TamaguiOptions } from '@tamagui/static'
import path from 'node:path'
import type { Plugin, ResolvedConfig, ViteDevServer } from 'vite'
import { normalizePath, type Environment } from 'vite'
import {
  Static,
  disableStatic,
  extractor,
  loadTamaguiBuildConfig,
  tamaguiOptions,
} from './loadTamagui'

import { createHash } from 'node:crypto'

export function tamaguiExtractPlugin(optionsIn?: Partial<TamaguiOptions>): Plugin {
  if (optionsIn?.disable) {
    return {
      name: 'tamagui-extract',
    }
  }

  const getHash = (input: string) => createHash('sha1').update(input).digest('base64')

  const clearCompilerCache = () => {
    memoryCache = {}
    cacheSize = 0
  }

  let memoryCache = {}
  let cacheSize = 0

  const cssMap = new Map<string, string>()

  let config: ResolvedConfig
  let server: ViteDevServer
  const virtualExt = `.tamagui.css`

  const getAbsoluteVirtualFileId = (filePath: string) => {
    if (filePath.startsWith(config.root)) {
      return filePath
    }
    return normalizePath(path.join(config.root, filePath))
  }

  function isVite6AndNotClient(environment?: Environment) {
    return environment?.name && environment.name !== 'client'
  }

  function isVite6Native(environment?: Environment) {
    return (
      environment?.name && (environment.name === 'ios' || environment.name === 'android')
    )
  }

  function invalidateModule(absoluteId: string) {
    if (!server) return

    const { moduleGraph } = server
    const modules = moduleGraph.getModulesByFile(absoluteId)

    if (modules) {
      for (const module of modules) {
        moduleGraph.invalidateModule(module)

        // Vite uses this timestamp to add `?t=` query string automatically for HMR.
        module.lastHMRTimestamp = module.lastInvalidationTimestamp || Date.now()
      }
    }
  }

  return {
    name: 'tamagui-extract',
    enforce: 'pre',

    configureServer(_server) {
      server = _server
    },

    async buildStart() {
      await loadTamaguiBuildConfig(optionsIn)
    },

    buildEnd() {
      extractor?.cleanupBeforeExit()
    },

    config(userConf) {
      userConf.optimizeDeps ||= {}
      userConf.optimizeDeps.include ||= []
      userConf.optimizeDeps.include.push('@tamagui/core/inject-styles')
    },

    async configResolved(resolvedConfig) {
      config = resolvedConfig
    },

    async resolveId(source) {
      if (isVite6Native(this.environment)) {
        return
      }

      if (
        tamaguiOptions?.disableServerOptimization &&
        isVite6AndNotClient(this.environment)
      ) {
        // only optimize on client - server should produce identical styles anyway!
        return
      }

      const [validId, query] = source.split('?')

      if (!validId.endsWith(virtualExt)) {
        return
      }

      // Absolute paths seem to occur often in monorepos, where files are
      // imported from outside the config root.
      const absoluteId = source.startsWith(config.root)
        ? source
        : getAbsoluteVirtualFileId(validId)

      // There should always be an entry in the `cssMap` here.
      // The only valid scenario for a missing one is if someone had written
      // a file in their app using the .tamagui.js/.tamagui.css extension
      if (cssMap.has(absoluteId)) {
        // Keep the original query string for HMR.
        return absoluteId + (query ? `?${query}` : '')
      }
    },

    /**
     * TODO
     *
     *   mainFields module:jsx breaks, so lets just have a mapping here
     *   where we load() and map it to the jsx path before transform
     *
     */

    async load(id) {
      if (disableStatic) {
        // only optimize on client - server should produce identical styles anyway!
        return
      }

      if (isVite6Native(this.environment)) {
        return
      }

      if (
        tamaguiOptions?.disableServerOptimization &&
        isVite6AndNotClient(this.environment)
      ) {
        return
      }

      const [validId] = id.split('?')
      return cssMap.get(validId)
    },

    transform: {
      order: 'pre',
      async handler(code, id) {
        if (disableStatic) {
          // only optimize on client - server should produce identical styles anyway!
          return
        }

        if (isVite6Native(this.environment)) {
          return
        }

        if (
          tamaguiOptions?.disableServerOptimization &&
          isVite6AndNotClient(this.environment)
        ) {
          return
        }

        const [validId] = id.split('?')
        if (!validId.endsWith('.tsx')) {
          return
        }

        const firstCommentIndex = code.indexOf('// ')
        const { shouldDisable, shouldPrintDebug } = Static!.getPragmaOptions({
          source: firstCommentIndex >= 0 ? code.slice(firstCommentIndex) : '',
          path: validId,
        })

        if (shouldPrintDebug) {
          console.trace(
            `Current file: ${id} in environment: ${this.environment?.name}, shouldDisable: ${shouldDisable}`
          )
          console.info(`\n\nOriginal source:\n${code}\n\n`)
        }

        if (shouldDisable) {
          return
        }

        const cacheEnv =
          this.environment.name === 'client' || this.environment.name === 'ssr'
            ? // same cache key for ssr and web since they are the same
              'web'
            : this.environment.name
        const cacheKey = getHash(`${cacheEnv}${code}${id}`)
        const cached = memoryCache[cacheKey]
        if (cached) {
          return cached
        }

        const extracted = await Static!.extractToClassNames({
          extractor: extractor!,
          source: code,
          sourcePath: validId,
          options: tamaguiOptions!,
          shouldPrintDebug,
        })

        if (!extracted) {
          return
        }

        const rootRelativeId = `${validId}${virtualExt}`
        const absoluteId = getAbsoluteVirtualFileId(rootRelativeId)

        let source = extracted.js

        if (extracted.styles) {
          this.addWatchFile(rootRelativeId)

          if (server && cssMap.has(absoluteId)) {
            invalidateModule(rootRelativeId)
          }

          source = `${source}\nimport "${rootRelativeId}";`
          cssMap.set(absoluteId, extracted.styles)
        }

        const codeOut = source.toString()
        const out = {
          code: codeOut,
          map: extracted.map,
        }

        cacheSize += codeOut.length
        // ~50Mb cache for recently compiler files
        if (cacheSize > 26214400) {
          clearCompilerCache()
        }
        memoryCache[cacheKey] = out

        return out
      },
    },
  }
}
