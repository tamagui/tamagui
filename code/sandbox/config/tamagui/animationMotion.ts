import { createAnimations } from '@tamagui/animations-motion'

export const animationsMotion = createAnimations({
  '75ms': {
    type: 'time',
    duration: 75,
  },
  '100ms': {
    type: 'time',
    duration: 100,
  },
  '200ms': {
    type: 'time',
    duration: 200,
  },
  superBouncy: {
    type: 'spring',
    damping: 5,
    mass: 0.7,
    stiffness: 200,
  },
  bouncy: {
    type: 'spring',
    damping: 9,
    mass: 0.9,
    stiffness: 150,
  },
  medium: {
    type: 'spring',
    damping: 15,
    stiffness: 120,
    mass: 1,
  },
  lazy: {
    type: 'spring',
    damping: 18,
    stiffness: 50,
  },
  slow: {
    type: 'spring',
    damping: 15,
    stiffness: 40,
  },
  quick: {
    type: 'spring',
    damping: 20,
    mass: 1.2,
    stiffness: 250,
  },
  tooltip: {
    type: 'spring',
    damping: 10,
    mass: 0.9,
    stiffness: 100,
  },
  quicker: {
    type: 'spring',
    damping: 20,
    mass: 0.7,
    stiffness: 250,
  },
  quickest: {
    type: 'spring',
    damping: 5,
    mass: 0.2,
    stiffness: 300,
  },
})
