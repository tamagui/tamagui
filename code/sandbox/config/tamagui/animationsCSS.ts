import { createAnimations } from '@tamagui/animations-css'

const smoothBezier = 'cubic-bezier(0.215, 0.610, 0.355, 1.000)'

export const animationsCSS = createAnimations({
  '75ms': 'ease-in 75ms',
  '100ms': 'ease-in 100ms',
  '200ms': 'ease-in 200ms',
  bouncy: 'ease-in 200ms',
  superBouncy: 'ease-in 500ms',
  lazy: 'ease-in 1000ms',
  medium: 'ease-in 300ms',
  slow: 'ease-in 500ms',
  quick: `${smoothBezier} 400ms`,
  quicker: `${smoothBezier} 300ms`,
  quickest: `${smoothBezier} 200ms`,
  tooltip: 'ease-in 400ms',
})
