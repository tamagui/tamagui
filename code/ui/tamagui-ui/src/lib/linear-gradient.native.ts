export * from 'expo-linear-gradient'
