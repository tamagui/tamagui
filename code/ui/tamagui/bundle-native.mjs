#!/usr/bin/env node

import { bundleNative } from '@tamagui/native-bundle'

await bundleNative({
  entry: 'src/index.ts',
  outDir: 'dist',
  fileName: 'native.cjs',
})
