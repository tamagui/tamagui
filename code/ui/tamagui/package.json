{
  "name": "tamagui",
  "version": "1.124.13",
  "description": "Style and UI for React (web and native) meet an optimizing compiler",
  "removeSideEffects": true,
  "alsoPublishAs_disabled": [
    "@tamagui/ui"
  ],
  "sideEffects": [
    "setup.js"
  ],
  "source": "src/index.ts",
  "types": "./types/index.d.ts",
  "main": "dist/cjs",
  "module": "dist/esm",
  "module:jsx": "dist/jsx",
  "tamagui": {
    "bundle.native": "./src/index.ts",
    "bundle.native.test": "./src/index.ts"
  },
  "files": [
    "src",
    "types",
    "dist",
    "linear-gradient",
    "native.js",
    "native.d.ts"
  ],
  "scripts": {
    "build": "tamagui-build",
    "watch": "tamagui-build --watch",
    "clean": "tamagui-build clean",
    "clean:build": "tamagui-build clean:build",
    "lint": "biome check src",
    "lint:fix": "biome check --write src",
    "check": "yarn check-circular-deps",
    "check-circular-deps": "npx madge --circular ./src/index.ts"
  },
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "react-native": {
        "import": "./dist/esm/index.native.js",
        "require": "./dist/cjs/index.native.js"
      },
      "types": "./types/index.d.ts",
      "import": "./dist/esm/index.mjs",
      "require": "./dist/cjs/index.cjs",
      "default": "./dist/cjs/index.native.js"
    },
    "./react-native-web": {
      "react-native": "./dist/cjs/react-native-web.native.js",
      "types": "./types/index.d.ts",
      "import": "./dist/esm/react-native-web.mjs",
      "require": "./dist/cjs/react-native-web.cjs"
    },
    "./web": {
      "react-native": {
        "import": "./dist/esm/index.native.js",
        "require": "./dist/cjs/index.native.js"
      },
      "types": "./types/index.d.ts",
      "import": "./dist/esm/index.mjs",
      "require": "./dist/cjs/index.cjs"
    },
    "./native": {
      "types": "./types/index.d.ts",
      "require": "./dist/native.js",
      "react-native": "./dist/native.native.js"
    },
    "./native-test": {
      "types": "./types/index.d.ts",
      "require": "./dist/test.cjs",
      "react-native": "./dist/test.native.js"
    },
    "./linear-gradient": {
      "react-native": {
        "import": "./dist/esm/linear-gradient.native.js",
        "require": "./dist/cjs/linear-gradient.native.js"
      },
      "types": "./types/linear-gradient.d.ts",
      "import": "./dist/esm/linear-gradient.mjs",
      "require": "./dist/cjs/linear-gradient.cjs"
    }
  },
  "dependencies": {
    "@tamagui/accordion": "workspace:*",
    "@tamagui/adapt": "workspace:*",
    "@tamagui/alert-dialog": "workspace:*",
    "@tamagui/animate-presence": "workspace:*",
    "@tamagui/avatar": "workspace:*",
    "@tamagui/button": "workspace:*",
    "@tamagui/card": "workspace:*",
    "@tamagui/checkbox": "workspace:*",
    "@tamagui/compose-refs": "workspace:*",
    "@tamagui/constants": "workspace:*",
    "@tamagui/core": "workspace:*",
    "@tamagui/create-context": "workspace:*",
    "@tamagui/dialog": "workspace:*",
    "@tamagui/elements": "workspace:*",
    "@tamagui/fake-react-native": "workspace:*",
    "@tamagui/focusable": "workspace:*",
    "@tamagui/font-size": "workspace:*",
    "@tamagui/form": "workspace:*",
    "@tamagui/get-button-sized": "workspace:*",
    "@tamagui/get-font-sized": "workspace:*",
    "@tamagui/get-token": "workspace:*",
    "@tamagui/group": "workspace:*",
    "@tamagui/helpers-tamagui": "workspace:*",
    "@tamagui/image": "workspace:*",
    "@tamagui/label": "workspace:*",
    "@tamagui/linear-gradient": "workspace:*",
    "@tamagui/list-item": "workspace:*",
    "@tamagui/polyfill-dev": "workspace:*",
    "@tamagui/popover": "workspace:*",
    "@tamagui/popper": "workspace:*",
    "@tamagui/portal": "workspace:*",
    "@tamagui/progress": "workspace:*",
    "@tamagui/radio-group": "workspace:*",
    "@tamagui/react-native-media-driver": "workspace:*",
    "@tamagui/scroll-view": "workspace:*",
    "@tamagui/select": "workspace:*",
    "@tamagui/separator": "workspace:*",
    "@tamagui/shapes": "workspace:*",
    "@tamagui/sheet": "workspace:*",
    "@tamagui/slider": "workspace:*",
    "@tamagui/stacks": "workspace:*",
    "@tamagui/switch": "workspace:*",
    "@tamagui/tabs": "workspace:*",
    "@tamagui/text": "workspace:*",
    "@tamagui/theme": "workspace:*",
    "@tamagui/toggle-group": "workspace:*",
    "@tamagui/tooltip": "workspace:*",
    "@tamagui/use-controllable-state": "workspace:*",
    "@tamagui/use-debounce": "workspace:*",
    "@tamagui/use-force-update": "workspace:*",
    "@tamagui/use-window-dimensions": "workspace:*",
    "@tamagui/visually-hidden": "workspace:*"
  },
  "devDependencies": {
    "@tamagui/build": "workspace:*",
    "react": "*",
    "react-native-web": "^0.19.0"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/tamagui/tamagui.git",
    "directory": "code/ui/tamagui"
  },
  "gitHead": "a49cc7ea6b93ba384e77a4880ae48ac4a5635c14",
  "peerDependencies": {
    "react": "*"
  }
}
