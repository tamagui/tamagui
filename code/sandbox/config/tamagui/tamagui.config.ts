import { defaultConfig } from '@tamagui/config/v4'
import { createTamagui } from 'tamagui'

export const config = createTamagui(defaultConfig)

export type Conf = typeof config

declare module 'tamagui' {
  interface TamaguiCustomConfig extends Conf {}

  interface TypeOverride {
    groupNames(): 'card' | 'other'
  }
}

export default config
