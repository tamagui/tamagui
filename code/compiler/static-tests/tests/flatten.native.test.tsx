import * as React from 'react'
import { describe, expect, test } from 'vitest'

import { extractForNative } from './lib/extract'

Error.stackTraceLimit = Number.Infinity
process.env.TAMAGUI_TARGET = 'native'

window['React'] = React

describe('flatten-tests', () => {
  test(`flattened without extra attributes`, async () => {
    const output = await extractForNative(`
      import { YStack } from 'tamagui'
      import { useMedia } from 'tamagui'
  
      export function Test(isLoading) {
        const media = useMedia()
        
        return (
          <YStack
            y={10}
            x={20}
            rotate="10deg"
          />
        )
      }
    `)

    expect(output?.code).toContain(`<__ReactNativeView style={_sheet["0"]} />`)
  })

  test('flattened media queries', async () => {
    const output = await extractForNative(`
      import { YStack } from 'tamagui'
      import { useMedia } from 'tamagui'
  
      export function Test(isLoading) {
        const media = useMedia()
        
        return (
          <YStack
            y={10}
            x={20}
            rotate="10deg"
            {...media.sm && {
              scale: 2,
              borderRadius: 10,
              backgroundColor: isLoading ? 'red' : 'blue'
            }}
          />
        )
      }
    `)

    const code = output?.code ?? ''

    expect(code).toMatchSnapshot()

    const startStr = `ReactNativeStyleSheet.create(`
    const start = code.indexOf(startStr)
    const end = code.indexOf(`});`)
    const defs = code.slice(start + startStr.length, end + 1)
    const sheetStyles = JSON.parse(defs)

    expect(sheetStyles['0']).toEqual({
      transform: [
        {
          translateY: 10,
        },
        {
          translateX: 20,
        },
        {
          rotate: '10deg',
        },
      ],
      flexDirection: 'column',
    })

    expect(sheetStyles['1']).toEqual({
      borderTopLeftRadius: 10,
      borderTopRightRadius: 10,
      borderBottomRightRadius: 10,
      borderBottomLeftRadius: 10,
      transform: [
        {
          scale: 2,
        },
      ],
    })

    expect(sheetStyles['3']).toEqual({
      backgroundColor: 'red',
    })

    expect(sheetStyles['5']).toEqual({
      backgroundColor: 'blue',
    })
  })

  test(`work with experimentalFlattenThemesOnNative`, async () => {
    const output = await extractForNative(`
      import { YStack } from 'tamagui'
  
      export function Test(isLoading) {
        return (
          <YStack
            y={10}
            x={20}
            rotate="10deg"
            backgroundColor="$background"
          />
        )
      }
    `)

    expect(output?.code).toMatchSnapshot()
  })

  test(`work with experimentalFlattenThemesOnNative + ternary`, async () => {
    const output = await extractForNative(`
      import { View } from 'tamagui'
  
      export function Test() {
        return (
          <View backgroundColor={showBackground ? '$color1' : '$color2'} />
        )
      }
    `)

    expect(output?.code).toMatchSnapshot()
  })

  // TODO make this work:
  // test.skip(`keeps style object a single object case 2`, async () => {
  //   const output = await extractForNative(`
  //     import { View } from 'tamagui'

  //     export function Test() {
  //       return (
  //         <View position="absolute" key={0} right="$2" top="$2" />
  //       )
  //     }
  //   `)

  //   // just one sheet
  //   expect(output?.code).toContain(`style={_sheet["0"]}`)
  // })
})
