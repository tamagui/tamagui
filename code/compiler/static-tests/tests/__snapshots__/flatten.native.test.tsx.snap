// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`flatten-tests > flattened media queries 1`] = `
"const __ReactNativeStyleSheet = require('react-native').StyleSheet;
const _sheet = __ReactNativeStyleSheet.create({
  "0": {
    "flexDirection": "column",
    "transform": [{
      "translateY": 10
    }, {
      "translateX": 20
    }, {
      "rotate": "10deg"
    }]
  },
  "1": {
    "borderTopLeftRadius": 10,
    "borderTopRightRadius": 10,
    "borderBottomRightRadius": 10,
    "borderBottomLeftRadius": 10,
    "transform": [{
      "scale": 2
    }]
  },
  "2": {},
  "3": {
    "backgroundColor": "red"
  },
  "4": {},
  "5": {
    "backgroundColor": "blue"
  },
  "6": {}
});
const __ReactUseMemo = require('react').useMemo;
const __withStableStyle = require('@tamagui/core')._withStableStyle;
const __ReactNativeView = require('react-native').View;
const __ReactNativeText = require('react-native').Text;
import { YStack } from 'tamagui';
import { useMedia } from 'tamagui';
export function Test(isLoading) {
  const media = useMedia();
  return <_ReactNativeViewWrapper expressions={[media.sm, media.sm && isLoading, media.sm && !isLoading]} />;
}
const _ReactNativeViewWrapper = __withStableStyle(__ReactNativeView, (theme, _expressions) => {
  return __ReactUseMemo(() => {
    return [_sheet["0"], _expressions[0] ? _sheet["1"] : _sheet["2"], _expressions[1] ? _sheet["3"] : _sheet["4"], _expressions[2] ? _sheet["5"] : _sheet["6"]];
  }, [..._expressions]);
});"
`;

exports[`flatten-tests > work with experimentalFlattenThemesOnNative + ternary 1`] = `
"const __ReactNativeStyleSheet = require('react-native').StyleSheet;
const _sheet = __ReactNativeStyleSheet.create({
  "0": {},
  "1": {}
});
const __ReactUseMemo = require('react').useMemo;
const __withStableStyle = require('@tamagui/core')._withStableStyle;
const __ReactNativeView = require('react-native').View;
const __ReactNativeText = require('react-native').Text;
import { View } from 'tamagui';
export function Test() {
  return <_ReactNativeViewWrapper expressions={[showBackground]} />;
}
const _ReactNativeViewWrapper = __withStableStyle(__ReactNativeView, (theme, _expressions) => {
  return __ReactUseMemo(() => {
    return [_sheet["0"], _sheet["1"], _expressions[0] ? {
      "backgroundColor": theme.color1.get()
    } : {
      "backgroundColor": theme.color2.get()
    }];
  }, [..._expressions]);
});"
`;

exports[`flatten-tests > work with experimentalFlattenThemesOnNative 1`] = `
"const __ReactNativeStyleSheet = require('react-native').StyleSheet;
const _sheet = __ReactNativeStyleSheet.create({
  "0": {
    "flexDirection": "column",
    "transform": [{
      "translateY": 10
    }, {
      "translateX": 20
    }, {
      "rotate": "10deg"
    }]
  }
});
const __ReactUseMemo = require('react').useMemo;
const __withStableStyle = require('@tamagui/core')._withStableStyle;
const __ReactNativeView = require('react-native').View;
const __ReactNativeText = require('react-native').Text;
import { YStack } from 'tamagui';
export function Test(isLoading) {
  return <_ReactNativeViewWrapper />;
}
const _ReactNativeViewWrapper = __withStableStyle(__ReactNativeView, (theme, _expressions) => {
  return __ReactUseMemo(() => {
    return [_sheet["0"], {
      "backgroundColor": theme.background.get()
    }];
  }, [..._expressions]);
});"
`;
